server/src/main/java/org/elasticsearch/index/codec/vectors/es910/ES910BinaryQuantizedVectorsWriter.java

```diff
@@ -0,0 +1,955 @@
+                float[] vector = flatFieldVectorsWriter.getVectors().get(i);
+                float magnitude = magnitudes.get(i);
+                for (int j = 0; j < vector.length; j++) {
+                    vector[j] /= magnitude;
+                }
+            }
+        }
+
+        @Override
+        public DocsWithFieldSet getDocsWithFieldSet() {
+            return flatFieldVectorsWriter.getDocsWithFieldSet();
+        }
+
+        @Override
+        public void finish() throws IOException {
+            if (finished) {
+                return;
+            }
+            assert flatFieldVectorsWriter.isFinished();
+            finished = true;
+        }
+
+        @Override
+        public boolean isFinished() {
+            return finished && flatFieldVectorsWriter.isFinished();
+        }
+
+        @Override
+        public void addValue(int docID, float[] vectorValue) throws IOException {
+            flatFieldVectorsWriter.addValue(docID, vectorValue);
+            if (fieldInfo.getVectorSimilarityFunction() == COSINE) {
+                float dp = VectorUtil.dotProduct(vectorValue, vectorValue);
+                float divisor = (float) Math.sqrt(dp);
+                magnitudes.add(divisor);
+                for (int i = 0; i < vectorValue.length; i++) {
+                    dimensionSums[i] += (vectorValue[i] / divisor);
+                }
+            } else {
+                for (int i = 0; i < vectorValue.length; i++) {
+                    dimensionSums[i] += vectorValue[i];
+                }
+            }
+        }
+
+        @Override
+        public float[] copyValue(float[] vectorValue) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public long ramBytesUsed() {
+            long size = SHALLOW_SIZE;
+            size += flatFieldVectorsWriter.ramBytesUsed();
+            size += magnitudes.ramBytesUsed();
+            return size;
+        }
+    }
+
+    // When accessing vectorValue method, targerOrd here means a row ordinal.
+    static class OffHeapBinarizedQueryVectorValues {
+        private final IndexInput slice;
+        private final int dimension;
+        private final int size;
+        protected final byte[] binaryValue;
+        protected final ByteBuffer byteBuffer;
+        private final int byteSize;
+        protected final float[] correctiveValues;
+        private int lastOrd = -1;
+        private int quantizedComponentSum;
+
+        OffHeapBinarizedQueryVectorValues(IndexInput data, int dimension, int size) {
+            this.slice = data;
+            this.dimension = dimension;
+            this.size = size;
+            // 4x the quantized binary dimensions
+            int binaryDimensions = (BQVectorUtils.discretize(dimension, 64) / 8) * BQSpaceUtils.B_QUERY;
+            this.byteBuffer = ByteBuffer.allocate(binaryDimensions);
+            this.binaryValue = byteBuffer.array();
+            // + 1 for the quantized sum
```

Comment:
I am not sure if this is used, but its incorrect as you aren't packing the bits, binaryDimensions is just dimensions.